rosidl_runtime_c/test/test_string_functions.cpp

```diff
@@ -0,0 +1,149 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "gtest/gtest.h"
+#include "rosidl_runtime_c/string.h"
+#include "rosidl_runtime_c/string_functions.h"
+
+TEST(string_functions, init_fini_empty_string) {
+  rosidl_runtime_c__String empty_string;
+  EXPECT_TRUE(rosidl_runtime_c__String__init(&empty_string));
+  EXPECT_EQ(empty_string.size, 0u);
+  EXPECT_EQ(empty_string.capacity, 1u);
+  EXPECT_EQ(empty_string.data[0], 0);
+
+  // Should just return without changing anything.
+  rosidl_runtime_c__String__fini(nullptr);
+
+  rosidl_runtime_c__String__fini(&empty_string);
+  EXPECT_EQ(empty_string.data, nullptr);
+  EXPECT_EQ(empty_string.size, 0u);
+  EXPECT_EQ(empty_string.capacity, 0u);
+}
+
+TEST(string_functions, init_fini_bad_data) {
+  EXPECT_FALSE(rosidl_runtime_c__String__init(nullptr));
+
+  rosidl_runtime_c__String bad_string;
+  bad_string.size = 1u;
+  bad_string.capacity = 1u;
+  bad_string.data = nullptr;
+
+  int expected_error_code = -1;
+#ifndef _WIN32
+  // While exit() takes an int, when a parent process waits for a child process to die, it only
+  // takes the lowest 8-bits.
+  // See Note: in https://pubs.opengroup.org/onlinepubs/009695399/functions/exit.html
+  expected_error_code &= 255;
+#endif
+  EXPECT_EXIT(
+    rosidl_runtime_c__String__fini(&bad_string),
+    ::testing::ExitedWithCode(expected_error_code),
+    "Unexpected condition: string size was non-zero for deallocated data! Exiting.");
+  bad_string.size = 0u;
+
+  EXPECT_EXIT(
+    rosidl_runtime_c__String__fini(&bad_string),
+    ::testing::ExitedWithCode(expected_error_code),
+    "Unexpected behavior: string capacity was non-zero for deallocated data! Exiting.");
+
+  char c = 0;
+  bad_string.data = &c;
+  bad_string.capacity = 0u;
+
+  EXPECT_EXIT(
+    rosidl_runtime_c__String__fini(&bad_string),
+    ::testing::ExitedWithCode(expected_error_code),
+    "Unexpected condition: string capacity was zero for allocated data! Exiting.");
+}
+
+TEST(string_functions, resize_assignn) {
```

Comment:
Are you supposed to be able to call `assign` on an uninitialized or fini'd instance? Unless I'm misreading the code, I think that might work. If it is allowed, we should probably test for that.

Reply:
Added a check for assign after fini'd. But this doesn't work before init because s->data is an uninitialized pointer, which results in a segfault when passed to realloc().